### PR TITLE
cross platform: add edge date cases support

### PR DESCRIPTION
**Pull request by obastemur**
This PR improves xplat date implementation with edge cases support.
Now test cases for Date passes on Ubuntu / Debian

 - tested against VSOFull and its test cases
 - no perf difference on Windows measured

Update:

Some details on what's been done with this update.

  - Linux implementation no longer converts from Windows SYSTEMTIME.
    ChakraCore's internal YMD structure was a good fit between tm and
    SYSTEMTIME. Besides, new implementation spends less time there.

  - Now xplat handles out of range year cases ( <1900 or > 2100) and
    Feb28/29 + 1 (see test file under Date)

  - ChakraCore's common date utilities consider the start year is epoch.
    Besides, system time and clock is based on epoch too. However, the C
    api that calculates the time zone offset, and DST in case of leap year
    depends on broken-out time starting from 1900.
    We handle the leap year differences.

  - On xplat we need exact date to grab right tm_zone information instead
    of caching approach on Windows. Windows API can bring back both DTS and
    STD abbrv with a single call.

### CODE REVIEW

**Comment by obastemur** [comment]
Ping @digitalinfinity @jianchun @curtisman

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 1]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
+    static inline bool IsLeap(const int year)
+    {
+        return (0 == (year & 3)) && (0 != (year % 100) || 0 == (year % 400));
     }
 
-    const WCHAR *Utility::GetDaylightName(size_t *nameLength) {
-        // We have an abbreviated standard or daylight name
-        // based on the date and time zone we are in.
-        return GetStandardName(nameLength);
+    static inline int NormalizeYMDYear(const int base_year)
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Can you add some comments on what this method is for? Why 2100 or 1900?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 2]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
 
-    #define updatePeriod 1000
-
-    const WCHAR *Utility::GetStandardName(size_t *nameLength) {
-        data.UpdateTimeZoneInfo();
-        *nameLength = data.standardZoneNameLength;
-        return data.standardZoneName;
+    static inline bool IsLeap(const int year)
+    {
+        return (0 == (year & 3)) && (0 != (year % 100) || 0 == (year % 400));
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Instead of year & 3, can you do year % 4? It makes it more readable and the compiler should take care of optimizing it to & 3

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 3]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
 
-            tzset();
-            localtime_r(&gtime, &ltime);
+        return year;
+    }
 
-            standardZoneNameLength = strlen(ltime.tm_zone);
+    static void YMD_TO_TM(const YMD *ymd, struct tm *time, bool *leap_added)
+    {
+        time->tm_year = NormalizeYMDYear(ymd->year);
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Based on the usage of this method, would YMDYearToTmYear (and the inverse) be better names for these methods?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 4]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
-        // It can jump forwards and backwards.
-
-        sysTime->wSecond = time->tm_sec % 60;
+        *length = strlen(tm_zone);
+
+#if defined(WCHAR_IS_CHAR16_T)
+        for(int i = 0; i < *length; i++) {
+            wstr[i] = (char16_t)tm_zone[i];
+        }
+#elif defined(WCHAR_IS_WCHAR_T)
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Can we get rid of this? Should be safe to assume that on non-Win32 platforms, wchar is char16?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 5]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
+        wstr[*length] = (WCHAR)0;
     }
 
-    #define MIN_ZERO(value) value < 0 ? 0 : value
+    const WCHAR *Utility::GetStandardName(size_t *nameLength, const DateTime::YMD *ymd) {
+        AssertMsg(ymd != NULL, "xplat needs DateTime::YMD is defined for this call");
+        struct tm time_tm;
+        bool leap_added;
+        YMD_TO_TM(ymd, &time_tm, &leap_added);
+        mktime(&time_tm); // get zone name for the given date
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Curious- does this use historic timezone information to get the timezone name? If you didn't pass in ymd, and instead called mktime with time_tm set to now, would you get the same answer as what the Windows version returns? Because it looks like the Windows version will return the timezone name based on the rules for the current time, whereas this might return it based on historic information for the passed in date?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 3 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: YMD doesn't have a special year though. These methods normalize the abnormal dates temporarily for system calls

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: I couldn't see the same optimization with O2, Os or O3

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 5 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: What if the date you give to mktime is in standard time while the actual Date.toString within DST? There is no such issue in Windows.

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 6]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
+        int milliseconds = local->time % 1000;
 
         tzset();
+        int offset;
         time_t utime = timegm(&local_tm);
-        mktime(&local_tm); // we just want the gmt_off, don't care the result
-        utime -= local_tm.tm_gmtoff; // reverse UTC
+
+        // mktime min year 1900
+        if (local_tm.tm_year < 1900)
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Nit: minor readability suggestion:
since local_tm is not used beyond this, you could probably just do the following:
if (local_tm.tm_year < 1900)
{
    local_tm.tm_year = 1900;
}

mktime(&local_tm);
offset = local_tm.tm_gmtoff;

What do you think?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 7]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
-        time->tm_hour = sysTime->wHour;
-        time->tm_min = sysTime->wMinute;
-        time->tm_sec = sysTime->wSecond;
+    const WCHAR *Utility::GetDaylightName(size_t *nameLength, const DateTime::YMD *ymd) {
+        // xplat only gets the actual zone name for the given date
+        return GetStandardName(nameLength, ymd);
     }
 
-    static void SysLocalToUtc(SYSTEMTIME *local, SYSTEMTIME *utc)
+    static void SysLocalToUtc(YMD *local, YMD *utc)
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Sys over here in SysLocalToUtc actually referes to SYSTEMTIME. Since you are now passing in YMD, you should change this to YmdLocalToUtc

**Comment by obastemur** [comment]
@digitalinfinity thanks for the review. PR is updated.

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp [hunk 8]
```diff
@@ -13,155 +13,219 @@ namespace PlatformAgnostic
...
         // put milliseconds back
-        local->wMilliseconds = utc->wMilliseconds;
-
-        isDaylightSavings = local_tm.tm_isdst;
+        local->time += milliseconds;
+
+        // ugly hack but;
+        // right in between dst pass
+        // we need mktime trick to get the correct dst
+        utc_tm.tm_isdst = 1;
```
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Instead of doing this, can you just use the timezone variable that is set by tzset?

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 2 above)
Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Wow- that's really surprising- ok, leave it as is then but add the above information as a comment in the code so that no one accidentally removes it.

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 8 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Actually we already get the above. However, that doesn't work with leap year dst pass and we don't know that moment until we trick the system. As mentioned with the comment, this is an ugly trick : (

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Windows implementation FLeap does the same. I have carried this method from there.

Comment (digitalinfinity) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: Can we share the method then?

**Comment by curtisman** [comment]
Overall it looks fine, but I didn't look at the logic too closely.

File: lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Linux/DateTime.cpp]: We could but keeping leap related arithmetic separate on xplat.

**Comment by obastemur** [comment]
@digitalinfinity @curtisman Thanks for the review
